aws_signing_helper/pkcs11_signer.go

```diff
@@ -284,44 +560,83 @@ func getManufacturerId(module *pkcs11.Ctx) (string, error) {
 	return info.ManufacturerID, nil
 }
 
+// Because of *course* we have to do this for ourselves.
+//
+// Create the DER-encoded SEQUENCE containing R and S:
+//
+//	Ecdsa-Sig-Value ::= SEQUENCE {
+//	  r                   INTEGER,
+//	  s                   INTEGER
+//	}
+//
+// This is defined in RFC3279 §2.2.3 as well as SEC.1.
+// I can't find anything which mandates DER but I've seen
+// OpenSSL refusing to verify it with indeterminate length.
+func encode_ecdsa_sig_value(signature []byte) (out []byte, err error) {
+	siglen := len(signature) / 2
+
+	return asn1.Marshal(struct {R *big.Int; S *big.Int}{
+		big.NewInt(0).SetBytes(signature[:siglen]),
+		big.NewInt(0).SetBytes(signature[siglen:])})
+}
+
 // Checks whether the private key and certificate are associated with each other
-func checkPrivateKeyMatchesCert(module *pkcs11.Ctx, session pkcs11.SessionHandle, pinPkcs11 string, privateKeyHandle pkcs11.ObjectHandle, certificate *x509.Certificate, manufacturerId string) bool {
+func checkPrivateKeyMatchesCert(module *pkcs11.Ctx, session pkcs11.SessionHandle, keyType uint, alwaysAuth uint, pinPkcs11 string, privateKeyHandle pkcs11.ObjectHandle, certificate *x509.Certificate, manufacturerId string) bool {
 	var digestSuffix []byte
+	var mechanism uint
 	publicKey := certificate.PublicKey
 	ecdsaPublicKey, isEcKey := publicKey.(*ecdsa.PublicKey)
 	if isEcKey {
 		digestSuffixArr := sha256.Sum256(append([]byte("IAM RA"), elliptic.Marshal(ecdsaPublicKey, ecdsaPublicKey.X, ecdsaPublicKey.Y)...))
 		digestSuffix = digestSuffixArr[:]
+		mechanism = pkcs11.CKM_ECDSA
+		if keyType != pkcs11.CKK_EC {
+			return false
+		}
 	}
 
 	rsaPublicKey, isRsaKey := publicKey.(*rsa.PublicKey)
 	if isRsaKey {
 		digestSuffixArr := sha256.Sum256(append([]byte("IAM RA"), x509.MarshalPKCS1PublicKey(rsaPublicKey)...))
 		digestSuffix = digestSuffixArr[:]
+		mechanism = pkcs11.CKM_SHA256_RSA_PKCS
+		if keyType != pkcs11.CKK_RSA {
+			return false
+		}
 	}
 	// "AWS Roles Anywhere Credential Helper PKCS11 Test" || PKCS11_TEST_VERSION ||
 	// MANUFACTURER_ID || SHA256("IAM RA" || PUBLIC_KEY_BYTE_ARRAY)
 	digest := "AWS Roles Anywhere Credential Helper PKCS11 Test" +
 		strconv.Itoa(int(PKCS11_TEST_VERSION)) + manufacturerId + string(digestSuffix)
 	digestBytes := []byte(digest)
 	hash := sha256.Sum256(digestBytes)
-
-	err := module.SignInit(session, []*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_SHA256_RSA_PKCS, nil)}, privateKeyHandle)
-	if err != nil {
-		return false
+	if isEcKey {
+		// For CKM_ECDSA we pass in a hash, not the plain message
+		digestBytes = hash[:]
 	}
 
-	err = module.Login(session, pkcs11.CKU_CONTEXT_SPECIFIC, pinPkcs11)
+	// XX: Why are we duplicating this code from our actual Sign() function?
```

Comment:
Yeah, that's my bad. I think the common functionality could be refactored into another method.

Reply:
I'd like to go a bit further; that `checkPrivateKeyMatchesCert()` method can be lifted out and used for checking that TPM keys match their cert too. And in fact shouldn't we also check that filesystem keys match *their* cert?